### PR TITLE
Example of learning application with PASEOS.

### LINKED ISSUE #106: Example of learning application with communications in PASEOS
# Feature
## Desired Behavior / Functionality
Showcase communications and learning using PASEOS

## What Needs to Be Done
Implement notebook

### PR DESCRIPTION
**Pull request by johanos1**
Please enter the commit message for your changes. Lines starting

# Description
Showcase communications and learning using PASEOS

Summary of changes
- New notebook

## Resolved Issues

- [ ] fixes #106 

## How Has This Been Tested?
Via simulation

## Related Pull Requests
-

### CODE REVIEW

File: examples/Learning_example/learning_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,2295 @@
+{
```
Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    import sys 
</pre><p>Could you add a markdown cell at the top explaining what this notebook does?</p><p> </p><p>Also need to mention that you need sklearn and torch in addition to regular paseos requirements.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <p>Ideally, I would explain here a bit that your are encapsulating an paseos instance together with a neural network and its training routines inside a node. And that it provides a few functions to train, update, etc.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    class Node:
</pre><p>So, I am wondering here how much of this class we should show and how much we can put in a file and forget. Maybe it would make sense to put this thing into a node.py?</p><p> </p><p>Overall, could you add docstrings for the methods? :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #18.            tx_bits = self.model_size()
</pre><p>what is tx? please spell out</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #20.            self.paseos.register_activity("Train", activity_function=self.train, power_consumption_in_watt=0)
</pre><p>maybe nice to make the power consumption a parameter you can specify when creating the node?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #36.        def add_node(self, node):
</pre><p>would rename, it's a bit confusing that the node can add a node. maybe "add_known_node"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #48.        def merge_model(self, rx_model):
</pre><p>what is rx? maybe spell out?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <p>a plot or picture of their orbits may be nice to illustrate?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    paseos_cfg.sim.time_multiplier = 200
</pre><p>I would add a short explanation what this does to the markdown cell above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    pos1 = [0, 10000000, 0]
</pre><p>Just computed pos, v for two sats in similar orbits as sentinel. Just use v1 for both sats if you want them to move in opposing directions.</p><p> </p><p>sat 1</p><p>r at ref. = [-6361974.5152151017, 52215.393130868499, -345258.19754751836]</p><p>v at ref. = [433.46690432313761, 1180.9649876192523, -7808.7670808152052]</p><p>sat 2</p><p>r at ref. = [6361974.5152151017, -52215.393130868615, 345258.19754751911]</p><p>v at ref. = [-433.46690432313858, -1180.9649876192523, 7808.7670808152052]</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.    # make the nodes aware of each other
</pre><p>maybe it would be nicer to do this during the simulation whenever they are in line of sight? similar to the constellation notebook where I update the known actors for each sat every iteration?</p><p> </p><p>Otherwise it may seem like you ignore line of sight </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <p>Could you add some text explaining the dataset and problem? :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    X1 = np.asarray(node1.model.train_dataloader.dataset.X)
</pre><p>Why is t here no training data at X=0 neither for node1 or node2? Won't that mess up test accuracy near X=0?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <p>As in the other plot, for paper I would suggest to:</p><p> </p><p>* dpi=150,300 (this should also fix the rather small text size right now)</p><p>* add legend on what the vertical lines are</p><p>* hypothetically you could also use orbital revolutions as x axis to have a more interpretable time axis? Would also fit other plots in the paper. Orbital period for the orbits I provided is 5061.492804584637 seconds.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    axes.plot(range(len(acc1)), acc1, label="node1")
</pre><p>in previous plot it was node 1 instead of node1</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    axes.legend()
</pre><p>I think you can add a lot of the output spam (like the &lt;matplotlib.legend.Legend at 0x14ac3ca30&gt; ) by adding ; at the of lines doing plotting etc. that should surpress it. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

**Review by gomezzz — CHANGES_REQUESTED** [review]
Added some suggestions for paper and example on https://app.reviewnb.com/aidotse/PASEOS/pull/107/ :v:

File: examples/Learning_example/learning_example.ipynb (hunk 1 above)
Comment (johanos1) [reply examples/Learning_example/learning_example.ipynb:1]: <p>We can add so we dont have gaps.</p>

[the remark this answers is not in the document]
Comment (johanos1) [reply examples/Learning_example/learning_example.ipynb:1]: <p>I added something similar now.</p>

Comment (johanos1) [reply examples/Learning_example/learning_example.ipynb:1]: <p>What is the easiest way to plot the orbits of an actor?</p>

**Comment by johanos1** [comment]
I think we need to choose better parameters for the state of charge to be interesting. Perhaps faster charging? Also, we don't see much of the benefits from collaborative learning as the satellites see each other so seldomly with this constellation...

File: examples/Learning_example/learning_example.ipynb (hunk 1 above)
Comment (gomezzz) [reply examples/Learning_example/learning_example.ipynb:1]: <p>I'd suggest to create orbits with get_constellation, then the user can on the fly change the orbits if they want and you can use the pykep planets directly to plot.</p><p> </p><pre class="ql-syntax" spellcheck="false">from get_constellation import get_constellation

altitude = 550 * 1000 # Constellation attitude above the Earth's ground

inclination = 30.0 # inclination of each orbital plane

nPlanes = 4 # the number of orbital planes (see linked wiki article)

nSats = 4 # the number of satellites per orbital plane

t0 = pk.epoch_from_string("2022-Oct-27 21:00:00") # the starting date of our simulation




# Compute orbits of LEO satellites

planet_list,sats_pos_and_v = get_constellation(altitude,inclination,nSats,nPlanes,t0)




fig = plt.figure(figsize=(6,6),dpi=100)

ax = plt.axes(projection='3d')

colors = sns.color_palette("Paired")

for i in range (nPlanes*nSats):

&nbsp; &nbsp; color_idx = i // nPlanes

&nbsp; &nbsp; pk.orbit_plots.plot_planet(planet_list[i],axes=ax,s=64,color=colors[color_idx])
</pre><p> </p><p>to get opposing directions you will need to multiply the v of the second sat by -1 but in the orbit plot direction isn't show so doesn't matter here.</p>

Comment (gomezzz) [reply examples/Learning_example/learning_example.ipynb:1]: <p>(PS: the current edition also returns orbital period)</p>

File: examples/Learning_example/learning_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,2221 @@
...
```
Comment (gomezzz) [inline examples/Learning_example/learning_example.ipynb:1]: <blockquote>aggregates thse with</blockquote><p>aggregate those</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/PASEOS/pull/107/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

**Comment by gomezzz** [comment]
> I think we need to choose better parameters for the state of charge to be interesting. Perhaps faster charging? 

@johanos1 Maybe use the same values as in the other notebook?

```py
# Device
ActorBuilder.set_power_devices(actor=sat_actor,battery_level_in_Ws=10000+np.random.rand()*90000,
                                   max_battery_level_in_Ws=100000,charging_rate_in_W=50)

# Power for activities
100.0,"Processing"
2.0, "Standby"
```

> Also, we don't see much of the benefits from collaborative learning as the satellites see each other so seldomly with this constellation...

This is trickier. So there are two options:
* Go lower altitude (not below 200km though): Then you have a shorter orbital period so they see other more frequently.
* Go higher altitude: Then you have longer periods of visibility between them but also longer periods of no comms as orbital period increases

**Review by gomezzz — APPROVED** [review]
